src/proxy/cpsubsystem/SessionlessSemaphoreProxy.ts

```diff
@@ -0,0 +1,184 @@
+    }
+
+    private doTryAcquire(permits: number, timeout: number): Promise<boolean> {
+        const invocationUid = UuidUtil.generate();
+        return this.getClusterWideThreadId()
+            .then((clusterWideThreadId) =>
+                this.encodeInvokeOnRandomTarget(
+                    SemaphoreAcquireCodec,
+                    this.groupId,
+                    this.objectName,
+                    NO_SESSION_ID,
+                    clusterWideThreadId,
+                    invocationUid,
+                    permits,
+                    Long.fromNumber(timeout)
+                )
+            )
+            .then((clientMessage) => {
+                const response = SemaphoreAcquireCodec.decodeResponse(clientMessage);
+                return response.response;
+            })
+            .catch((err) => {
+                if (err instanceof WaitKeyCancelledError) {
+                    throw new IllegalStateError('Semaphore[' + this.objectName
+                        + '] not acquired because the acquire call on the CP group was cancelled.');
+                }
+                throw err;
+            });
+    }
+
+    release(permits = 1): Promise<void> {
+        assertNonNegativeNumber(permits);
+
+        const invocationUid = UuidUtil.generate();
+        return this.getClusterWideThreadId()
+            .then((clusterWideThreadId) =>
+                this.encodeInvokeOnRandomTarget(
+                    SemaphoreReleaseCodec,
+                    this.groupId,
+                    this.objectName,
+                    NO_SESSION_ID,
+                    clusterWideThreadId,
+                    invocationUid,
+                    permits
+                )
+            )
+            .then();
+    }
+
+    availablePermits(): Promise<number> {
+        return this.encodeInvokeOnRandomTarget(SemaphoreAvailablePermitsCodec, this.groupId, this.objectName)
+            .then((clientMessage) => {
+                const response = SemaphoreAvailablePermitsCodec.decodeResponse(clientMessage);
+                return response.response;
+            });
+    }
+
+    drainPermits(): Promise<number> {
+        const invocationUid = UuidUtil.generate();
+        return this.getClusterWideThreadId()
+            .then((clusterWideThreadId) =>
+                this.encodeInvokeOnRandomTarget(
+                    SemaphoreDrainCodec,
+                    this.groupId,
+                    this.objectName,
+                    NO_SESSION_ID,
+                    clusterWideThreadId,
+                    invocationUid
+                )
+            )
+            .then((clientMessage) => {
+                const response = SemaphoreDrainCodec.decodeResponse(clientMessage);
+                return response.response;
+            });
+    }
+
+    reducePermits(reduction: number): Promise<void> {
+        assertNonNegativeNumber(reduction);
+        if (reduction === 0) {
+            return;
```

Comment:
Should we return promise here?

Reply:
Addressed in c5b06afb1713e0b71a30e3921b7caba63dbbfb18